File: simple-stack-example-mvvm-fragments/src/main/java/com/example/android/architecture/blueprints/todoapp/application/BaseKey.java
package com.example.android.architecture.blueprints.todoapp.application;

import android.content.res.Resources;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;

import com.example.android.architecture.blueprints.todoapp.core.viewmodels.ViewModelLifecycleHelper;

/**
 * Created by Zhuinden on 2017.07.26..
 */

public abstract class BaseKey<T>
        implements Parcelable {
    protected abstract boolean isFabVisible();

    protected abstract void setupFab(Fragment fragment, FloatingActionButton fab);

    public String getFragmentTag() {
        return toString();
    }

    public String getViewModelTag() {
        return getFragmentTag() + "_VIEW_MODEL";
    }

    public abstract ViewModelLifecycleHelper.ViewModelCreator<T> getViewModelCreator();

    public final BaseFragment newFragment() {
        BaseFragment fragment = createFragment();
        Bundle bundle = fragment.getArguments();
        if(bundle == null) {
            bundle = new Bundle();
        }
        bundle.putParcelable("KEY", this);
        fragment.setArguments(bundle);
        return fragment;
    }

    protected abstract BaseFragment createFragment();

    @Nullable
    public String title(Resources resources) {
        return null;
    }

    public abstract int navigationViewId();

    public abstract boolean shouldShowUp();
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/core/navigation/BaseKey.kt
package com.zhuinden.simplestackdemoexamplefragments.core.navigation

import android.os.Bundle
import android.support.v4.app.Fragment

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

abstract class BaseKey : FragmentKey {
    override fun newFragment(): Fragment = createFragment().also { fragment ->
        fragment.arguments = (fragment.arguments ?: Bundle()).also { bundle ->
            bundle.putParcelable(BaseFragment.KEY_TAG, this)
        }
    }

    protected abstract fun createFragment(): Fragment

    override val fragmentTag: String
        get() = toString()
}
